src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteCache.java

```diff
@@ -248,50 +256,149 @@ public void ensureInputsPresent(
     }
     resourceName += "blobs/" + digestUtil.toString(digest);
 
-    @Nullable
-    HashingOutputStream hashOut =
-        options.remoteVerifyDownloads ? digestUtil.newHashingOutputStream(out) : null;
+    @Nullable Supplier<HashCode> hashSupplier = null;
+    if (options.remoteVerifyDownloads) {
+      HashingOutputStream hashOut = digestUtil.newHashingOutputStream(out);
+      hashSupplier = hashOut::hash;
+      out = hashOut;
+    }
+
     SettableFuture<Void> outerF = SettableFuture.create();
+    Futures.addCallback(
+        downloadBlob(resourceName, digest, out, hashSupplier),
+        new FutureCallback<Void>() {
+          @Override
+          public void onSuccess(Void result) {
+            outerF.set(null);
+          }
+
+          @Override
+          public void onFailure(Throwable t) {
+            outerF.setException(t);
+          }
+        },
+        Context.current().fixedContextExecutor(MoreExecutors.directExecutor()));
```

Comment:
currentContextExecutor(...)?

Reply:
Negative. currentContextExecutor will take the threads context _from the caller of execute_, not from the call of the method. In this case that is a scheduled executor thread for the retrier. I spent hours trying to figure out how currentContextExecutor was doing me wrong before I realized how. It has all the problems of directExecutor with a confusing name

Reply:
Negative. currentContextExecutor will take the threads context _from the caller of execute_, not from the call of the method. In this case that is a scheduled executor thread for the retrier. I spent hours trying to figure out how currentContextExecutor was doing me wrong before I realized how. It has all the problems of directExecutor with a confusing name